File: simulator/src/java/darkness/simulator/CommandSocket.java
package darkness.simulator;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.net.Socket;

public class CommandSocket extends Thread {
	private final PgmPlayer pgmPlayer;

	private static final String OVERLAY_PREFIX = "overlay:";

	public CommandSocket(PgmPlayer pgmPlayer) {
		this.pgmPlayer = pgmPlayer;
		setDaemon(true);
	}

	@Override
	public void run() {
		System.out.println("Socket running");
		ServerSocket serverSocket = null;
		try {
			serverSocket = new ServerSocket(13370);
			System.out.println("Listening...");
			while (true) {
				Socket socket = null;
				BufferedReader reader = null;
				try {
					socket = serverSocket.accept();
					reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
					System.out.println("Connected");
					String line;
					while ((line = reader.readLine()) != null) {
						System.out.println(line);
						if (line.startsWith(OVERLAY_PREFIX)) {
							String fileName = line.substring(OVERLAY_PREFIX.length());
							//TODO: Validate fileName against a whitelist to prevent reading of arbitrary files
							PgmReader pgmReader = new PgmReader(fileName);
							try {
								pgmReader.read();
								pgmPlayer.addOverlay(pgmReader);
							} catch (ParseException e) {
								System.out.println(String.format("Exception when parsing PGM file '%s': %s", fileName, e));
							} catch (IOException e) {
								System.out.println(String.format("Exception when reading PGM file '%s': %s", fileName, e));
							}
						}
					}
					System.out.println("Disconnected");
				} finally {
					if (reader != null) {
						reader.close();
					}
					if (socket != null) {
						socket.close();
					}
				}
			}
		} catch (IOException e) {
			throw new RuntimeException(e);
		} finally {
			if (serverSocket != null) {
				try {
					serverSocket.close();
				} catch (IOException e) {
					System.out.println("Exception when closing server socket: " + e);
				}
			}
		}
	}
}

File: simulator/src/kotlin/darkness/simulator/CommandSocket.kt
package darkness.simulator

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.ServerSocket
import java.net.Socket

class CommandSocket(private val pgmPlayer: PgmPlayer) : Thread() {
    private val overlayPrefix = "overlay:"

    init {
        isDaemon = true
    }

    override fun run() {
        println("Socket running")
        var serverSocket: ServerSocket? = null
        try {
            serverSocket = ServerSocket(13370)
            println("Listening...")
            while (true) {
                var socket: Socket? = null
                var reader: BufferedReader? = null
                try {
                    socket = serverSocket.accept()!!
                    reader = BufferedReader(InputStreamReader(socket.getInputStream()))
                    println("Connected")
                    reader.forEachLine {line ->
                        println(line)
                        if (line.startsWith(overlayPrefix)) {
                            val fileName = line.substring(overlayPrefix.length)
                            //TODO: Validate fileName against a whitelist to prevent reading of arbitrary files
                            val pgmReader = PgmReader(fileName)
                            try {
                                pgmReader.read()
                                pgmPlayer.addOverlay(pgmReader)
                            } catch (e: ParseException) {
                                println(String.format("Exception when parsing PGM file '%s': %s", fileName, e))
                            } catch (e: IOException) {
                                println(String.format("Exception when reading PGM file '%s': %s", fileName, e))
                            }

                        }
                    }
                    println("Disconnected")
                } finally {
                    reader?.close()
                    socket?.close()
                }
            }
        } finally {
            if (serverSocket != null) {
                try {
                    serverSocket.close()
                } catch (e: IOException) {
                    println("Exception when closing server socket: $e")
                }
            }
        }
    }
}
